source/server/hot_restarting_child.cc

```diff
@@ -0,0 +1,163 @@
+  wrapped_request.mutable_request()->mutable_pass_listen_socket()->set_address(address);
+  sendHotRestartMessage(parent_address_, wrapped_request);
+
+  std::unique_ptr<HotRestartMessage> wrapped_reply = receiveHotRestartMessage(Blocking::Yes);
+  if (!replyIsExpectedType(wrapped_reply.get(), HotRestartMessage::Reply::kPassListenSocket)) {
+    return -1;
+  }
+  return wrapped_reply->reply().pass_listen_socket().fd();
+}
+
+std::unique_ptr<HotRestartMessage> HotRestartingChild::getParentStats() {
+  if (restart_epoch_ == 0 || parent_terminated_) {
+    return nullptr;
+  }
+
+  HotRestartMessage wrapped_request;
+  wrapped_request.mutable_request()->mutable_stats();
+  sendHotRestartMessage(parent_address_, wrapped_request);
+
+  std::unique_ptr<HotRestartMessage> wrapped_reply = receiveHotRestartMessage(Blocking::Yes);
+  RELEASE_ASSERT(replyIsExpectedType(wrapped_reply.get(), HotRestartMessage::Reply::kStats),
+                 "Hot restart parent did not respond as expected to get stats request.");
+  return wrapped_reply;
+}
+
+void HotRestartingChild::drainParentListeners() {
+  if (restart_epoch_ == 0 || parent_terminated_) {
+    return;
+  }
+  // No reply expected.
+  HotRestartMessage wrapped_request;
+  wrapped_request.mutable_request()->mutable_drain_listeners();
+  sendHotRestartMessage(parent_address_, wrapped_request);
+}
+
+void HotRestartingChild::shutdownParentAdmin(HotRestart::ShutdownParentAdminInfo& info) {
+  if (restart_epoch_ == 0 || parent_terminated_) {
+    return;
+  }
+
+  HotRestartMessage wrapped_request;
+  wrapped_request.mutable_request()->mutable_shutdown_admin();
+  sendHotRestartMessage(parent_address_, wrapped_request);
+
+  std::unique_ptr<HotRestartMessage> wrapped_reply = receiveHotRestartMessage(Blocking::Yes);
+  RELEASE_ASSERT(replyIsExpectedType(wrapped_reply.get(), HotRestartMessage::Reply::kShutdownAdmin),
+                 "Hot restart parent did not respond as expected to ShutdownParentAdmin.");
+  info.original_start_time_ = wrapped_reply->reply().shutdown_admin().original_start_time();
+}
+
+void HotRestartingChild::terminateParent() {
+  if (restart_epoch_ == 0 || parent_terminated_) {
+    return;
+  }
+  HotRestartMessage wrapped_request;
+  wrapped_request.mutable_request()->mutable_terminate();
+  sendHotRestartMessage(parent_address_, wrapped_request);
+  parent_terminated_ = true;
+  // Once setting parent_terminated_ == true, we can send no more hot restart RPCs, and therefore
+  // receive no more responses, including stats. So, now safe to forget our stat transferral state.
+  parent_counter_values_.clear();
+  parent_gauge_values_.clear();
+}
+
+void HotRestartingChild::mergeParentStats(Stats::Store& stats_store,
+                                          const HotRestartMessage::Reply::Stats& stats_proto) {
+  // Map from stat name *substrings* to special logic to use for combining stat values.
+  const std::unordered_map<std::string, CombineLogic> combine_logic_exceptions{
```

Comment:
Thoughts on moving this to a dedicated .h or .cc file so it's obvious where to find it and we can potentially refer to that file in other comments to guide folks when adding new stats that might need special merging logic?

Reply:
Also can you add a comment for every stat here as to how/why you arrived at the combination setting? I think it would be useful for future readers and for review.

Reply:
Ping on this once we agree on the other things.

Reply:
Done.